### ISSUE TITLE
Dequeue op loses TensorShape

### ISSUE THREAD

**Issue by xiayandi** [tensorflow/tensorflow#4898]
I am not able to find related issues online.

Operating System:
Ubuntu, Mac OSX

Installed version of CUDA and cuDNN: 
CUDA 7.5, cuDNN v5.1

If installed from binary pip package, provide:
1. A link to the pip package you installed:
`TF_BINARY_URL=https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.11.0rc0-cp27-none-linux_x86_64.whl`
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.
`I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcublas.so locally`
`I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcudnn.so locally`
`I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcufft.so locally`
`I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcuda.so.1 locally`
`I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcurand.so locally`
`0.11.0rc0`

To reproduce the problem:

`a = tf.placeholder(dtype=tf.int32, shape=[None, 10])`
`print a.get_shape()`
`queue = tf.FIFOQueue(20, dtypes=[tf.int32])`
`queue.enqueue([a])`
`b = queue.dequeue()`
`print b.get_shape()`

The first print output <?, 10>, which is OK, and there is a TensorShape for a.
But the second print output is \<unknown\>, and there is no TensorShape for b.

TensorShape information is lost after dequeue operation.

And this TensorShape is required in LSTM cell class.

Can we add this TensorShape information back to the dequeued tensors?

Thanks.

**Comment by ebrevdo** [comment]
Have you tried tf.train.batch?

**Comment by xiayandi** [comment]
@ebrevdo Thanks for the reply. I plan to use this FIFOqueue for the input in my evaluation graph. And I don't want to specify batch_size for my evaluation graph.

And yes, I have tf.train.batch in my training graph. And it works fine.

Thanks.

**Comment by ebrevdo** [comment]
If you're reading one element at a time, why are you using a queue?

On Oct 11, 2016 1:36 PM, "Yandi Xia" <notifications@github.com> wrote:

@ebrevdo <https://github.com/ebrevdo> Thanks for the reply. I plan to use
this FIFOqueue for the input in my evaluation graph. And I don't want to
specify batch_size for my evaluation graph.

And yes, I have tf.train.batch in my training graph. And it works fine.

Thanks.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tensorflow/tensorflow/issues/4898#issuecomment-253037713>,
or mute the thread

**Comment by xiayandi** [comment]
@ebrevdo 
I don't specify batch size in the placeholder.
I enqueue a batch of instances into the queue (variable size of batches) each time.
And each dequeue will return a batch of data.

Thanks.

**Comment by ebrevdo** [comment]
Isn't that equivalent to passing enqueue_many=True to tf.train.batch?

**Comment by xiayandi** [comment]
@ebrevdo 
I don't think so. 

even if using tf.train.batch with enqueue_many=True, I think a fixed batch_size need to be set, right? I really need a variable batch_size here.

Further more, I use this input method for evaluation graph. And the graph will take in-memory input, rather than TFRecord or CSV files. Therefore, I don't think tf.train.batch can solve this problem.

Thank you very much!

**Comment by ebrevdo** [comment]
The batch_size is the batch size for the output; and it can be a scalar
Tensor; i.e., you can use a Placeholder for it.  tf.train.batch really just
uses a Queue underneath, it's agnostic to how you feed data.

Either way, if you're taking in-memory input, I still don't understand how
you're planning to use a queue.  Do you plan to feed input directly in one
python thread, and have the eval running in a separate thread?

**Comment by xiayandi** [comment]
@ebrevdo 

It is good to know that tf.train.batch can take placeholder as input. Thanks.

"Do you plan to feed input directly in one
python thread, and have the eval running in a separate thread?"
Yes, I create a thread for enqueue, and the input of the evaluation graph is the result of dequeue. After all the data in test set is enqueued, the thread halts and waits for the evaluation to be done.
This is how TensorFlow FIFOQueue works, right? The producer/consumer style?

Thanks

**Comment by ebrevdo** [comment]
Yep it is.

**Comment by xiayandi** [comment]
I found the solution. Actually there is a shapes argument in tf.FIFOQueue(). Just specify the shapes there, and the TensorShape info will be kept in the tensor objects.

Thanks @ebrevdo for your help and discussion!

**Comment by xiayandi** [comment]
No, actually tf.FIFOQueue() can't take shapes like [None, 10]. Every dimension needs to be specified. 

Is there any way that I can use FIFOQueue with variable batch_size without losing TensorShape information?

Thanks

**Comment by ebrevdo** [comment]
When you use FIFOQueue, you don't pass in the batch shape, only the
remainder.  For example, if you pass shapes like [None, 10] to
queue.enqueue_many(), you only use shape value [10] for that component.

**Comment by xiayandi** [comment]
Thanks @ebrevdo!

I just found a work around to this problem. Just using set_shape() to set shape of the dequeued tensor back.
For the above example, just use: 
`b.set_shape([None, 10])`

And then b will have a known TensorShape attribute.
